### PR TITLE
implement local origin errors

### LINKED ISSUE #32645: add support for local origin errors in outlier detection
Envoy supports splitting of local origin/remote errors while deciding a host to be outlier.  https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/cluster/v3/outlier_detection.proto#config-cluster-v3-outlierdetection.

This is useful in certain cases where we want to just rely on local failures to remove from load balancer pool.

The feature request is to add to Destination rule API.
     split_external_local_origin_errors
     consecutive_local_origin_failure
     enforcing_consecutive_local_origin_failure

### PR DESCRIPTION
**Pull request by ramaraochavali**
Implements istio/api#1989
Fixes #32645

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 1]
```diff
@@ -605,6 +605,18 @@ func applyOutlierDetection(c *cluster.Cluster, outlier *networking.OutlierDetect
 		out.MaxEjectionPercent = &wrappers.UInt32Value{Value: uint32(outlier.MaxEjectionPercent)}
 	}
 
+	if outlier.SplitExternalLocalOriginErrors {
+		out.SplitExternalLocalOriginErrors = true
+		if outlier.ConsecutiveLocalOriginFailures.GetValue() > 0 {
+			out.ConsecutiveLocalOriginFailure = &wrappers.UInt32Value{Value: outlier.ConsecutiveLocalOriginFailures.Value}
+		}
+		// When split external origin errors is set to true, we should disable consecutive 5xx and gateways.
+		out.EnforcingConsecutive_5Xx = &wrappers.UInt32Value{Value: 0}
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:614]: is it not valid to have both set? I may want 5 errors for local origin and 10 for remote?

If not, I think we should add this to validation, why not just reject it?

Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:614]: >  I may want 5 errors for local origin and 10 for remote?
Yes. We can do that. Actually above if we set to 0 we are ensuring that EnforcingConsecutive_5Xx is set to 0. So this is not needed. I will remove it.

**Comment by ramaraochavali** [comment]
@howardjohn changed PTAL

File: pilot/pkg/networking/core/v1alpha3/cluster.go [hunk 2]
```diff
@@ -605,6 +605,15 @@ func applyOutlierDetection(c *cluster.Cluster, outlier *networking.OutlierDetect
 		out.MaxEjectionPercent = &wrappers.UInt32Value{Value: uint32(outlier.MaxEjectionPercent)}
 	}
 
+	if outlier.SplitExternalLocalOriginErrors {
+		out.SplitExternalLocalOriginErrors = true
+		if outlier.ConsecutiveLocalOriginFailures.GetValue() > 0 {
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/cluster.go:610]: can/should we reject in validation webhook, SplitExternalLocalOriginErrors=false and ConsecutiveLocalOriginFailures is set?

**Review by howardjohn — APPROVED** [review]
can we make sure the validation logic makes sense for these changes in a followup?

File: pilot/pkg/networking/core/v1alpha3/cluster.go (hunk 2 above)
Comment (ramaraochavali) [reply pilot/pkg/networking/core/v1alpha3/cluster.go:610]: Yes. I will add validation logic

**Pull request merged**